src/platform/Darwin/system/SystemLayerImplDispatchSockets.mm

```diff
@@ -0,0 +1,293 @@
+        VerifyOrReturnError(nullptr != watch, CHIP_ERROR_INVALID_ARGUMENT);
+
+        watch->mCallback = callback;
+        watch->mCallbackData = data;
+        return CHIP_NO_ERROR;
+    }
+
+    CHIP_ERROR LayerImplDispatch::RequestCallback(SocketWatchToken token, SocketEventFlags flag)
+    {
+        __auto_type * watch = reinterpret_cast<SocketWatch *>(token);
+        VerifyOrReturnError(nullptr != watch, CHIP_ERROR_INVALID_ARGUMENT);
+
+        watch->mPendingIO.Set(flag);
+
+        __auto_type & source = (flag == SocketEventFlags::kRead) ? watch->mRdSource : watch->mWrSource;
+        __auto_type sourceType = (flag == SocketEventFlags::kRead) ? DISPATCH_SOURCE_TYPE_READ : DISPATCH_SOURCE_TYPE_WRITE;
+
+        VerifyOrReturnError(nullptr == source, CHIP_NO_ERROR);
+
+        // First time requesting callback for those events: install a dispatch source
+        __auto_type dispatchQueue = GetDispatchQueue();
+
+#if CONFIG_BUILD_FOR_HOST_UNIT_TEST
+        // Note: if no dispatch queue is available, callbacks most probably will not work, unless,
+        //       as in some tests from a test-specific local loop, the select based event handling is invoked.
+        VerifyOrReturnError(nullptr != dispatchQueue, CHIP_NO_ERROR);
+#else
+        VerifyOrDie(nullptr != dispatchQueue);
+#endif
+
+        source = dispatch_source_create(sourceType, static_cast<uintptr_t>(watch->mFD), 0, dispatchQueue);
+        VerifyOrReturnError(nullptr != source, CHIP_ERROR_NO_MEMORY);
+
+        dispatch_source_set_event_handler(source, ^{
+            if (watch->mPendingIO.Has(flag) && watch->mCallback != nullptr) {
+                SocketEvents events;
+                events.Set(flag);
+                watch->mCallback(events, watch->mCallbackData);
+            }
+        });
+        // only now we are sure the source exists and can become active
+        dispatch_activate(source);
+
+        return CHIP_NO_ERROR;
+    }
+
+    CHIP_ERROR LayerImplDispatch::ClearCallback(SocketWatchToken token, SocketEventFlags flag)
+    {
+        __auto_type * watch = reinterpret_cast<SocketWatch *>(token);
+        VerifyOrReturnError(nullptr != watch, CHIP_ERROR_INVALID_ARGUMENT);
+
+        watch->mPendingIO.Clear(flag);
+
+        return CHIP_NO_ERROR;
+    }
+
+    CHIP_ERROR LayerImplDispatch::RequestCallbackOnPendingRead(SocketWatchToken token)
+    {
+        return RequestCallback(token, SocketEventFlags::kRead);
+    }
+
+    CHIP_ERROR LayerImplDispatch::RequestCallbackOnPendingWrite(SocketWatchToken token)
+    {
+        return RequestCallback(token, SocketEventFlags::kWrite);
+    }
+
+    CHIP_ERROR LayerImplDispatch::ClearCallbackOnPendingRead(SocketWatchToken token)
+    {
+        return ClearCallback(token, SocketEventFlags::kRead);
+    }
+
+    CHIP_ERROR LayerImplDispatch::ClearCallbackOnPendingWrite(SocketWatchToken token)
+    {
+        return ClearCallback(token, SocketEventFlags::kWrite);
+    }
+
+#if !CONFIG_BUILD_FOR_HOST_UNIT_TEST
```

Comment:
This interaction with CONFIG_BUILD_FOR_HOST_UNIT_TEST _definitely_ needs to be documented.